iceoryx_examples/icedelivery/iox_publisher_typed_experimental.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/experimental/popo/publisher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+#include <iostream>
+#include <chrono>
+#include <thread>
+
+struct Position {
+    Position(double_t x, double_t y, double_t z) : x(x), y(y), z(z)
+    {};
+    double_t x = 0.0;
+    double_t y = 0.0;
+    double_t z = 0.0;
+};
+
+bool killswitch = false;
+
+static void sigHandler(int f_sig[[gnu::unused]])
+{
+    // caught SIGINT, now exit gracefully
+    killswitch = true;
+}
+
+void getVehiclePosition(Position* allocation, uint64_t multiplier)
+{
+    new (allocation) Position(1111.1111 * multiplier, 1111.1111 * multiplier, 1111.1111* multiplier);
+}
+
+int main(int argc, char *argv[])
+{
+    iox::runtime::PoshRuntime::getInstance("/iox-ex-publisher-modern");
+
+    auto typedPublisher = iox::popo::TypedPublisher<Position>({"Odometry", "Position", "Vehicle"});
+    typedPublisher.offer();
+
+    float_t ct = 0.0;
+    while(!killswitch)
+    {
+        ++ct;
+
+        // Retrieve a typed sample from shared memory.
+        // Sample can be held until ready to publish.
+        auto result = typedPublisher.loan();
+        if(!result.has_error())
+        {
+            auto& sample = result.get_value();
+            auto val = sample.get();
+            val->x = ct * 1.1;
+            val->y = ct * 1.1;
+            val->z = ct * 1.1;
+            typedPublisher.publish(sample);
+        }
+
+        // Retrieve a sample and provide logic to immediately populate and publish via a lambda.
+        typedPublisher.loan()
+            .and_then([&](iox::popo::Sample<Position>& sample){
+                auto allocation = sample.get();
+                // Do some stuff leading to eventually generating the data in the provided sample's shared memory...
+                new (allocation) Position(ct * 11.11, ct * 11.11, ct * 11.11);
+                // ...then publish the bytes
+                sample.publish();
```

Comment:
The subscriber doesn't receive samples. This can be reassessed. The options seem to be:
* Have different Sample classes for publisher and subscriber
* Publisher uses the Sample class whereas the subscriber works directly with unique_ptrs as the convenience methods provided by the Sample are not relevant there

Reply:
Scratch that, after some consideration I've updated the API so both publishers and subscribers work with `Sample`s.
Publishers work with a specialized version of `Sample` which has a publish method.
Subscribers receive samples that do not offer this method.
Both types of samples manage the loaned memory (via smart pointer) and additionally both contain a pointer to the `ChunkHeader`, should the user need it.